test/adapters/CharmAdapter.js

```diff
@@ -0,0 +1,734 @@
+        purchaseAmount,
+        strikeIndex,
+        maxCost,
+      } = params;
+
+      this.oTokenAddress = oTokenAddress;
+      this.underlying = underlying;
+      this.strikeAsset = strikeAsset;
+      this.collateralAsset = collateralAsset;
+      this.strikePrice = strikePrice;
+      this.expiry = expiry;
+      this.optionType = optionType;
+      this.purchaseAmount = purchaseAmount;
+      this.strikeIndex = strikeIndex;
+      this.paymentToken = this.optionType == PUT_OPTION_TYPE ? this.strikeAsset : this.underlying;
+
+      this.oToken = await ethers.getContractAt("IERC20", oTokenAddress);
+
+      this.optionViews = await ethers.getContractAt("IOptionViews", CHARM_OPTION_VIEWS);
+      this.market = await (await ethers.getContractAt("IOptionToken", oTokenAddress)).market();
+      this.marketContract = await (await ethers.getContractAt("IOptionMarket", this.market));
+      this.donor = "0x875abe6F1E2Aba07bED4A3234d8555A0d7656d12";
+
+      this.premium = await this.optionViews.getBuyOptionCost(this.market, this.collateralAsset == ONE_ADDRESS ? true : false, this.strikeIndex, this.purchaseAmount);
+      this.maxCost = this.maxCost || parseEther("9999999999");
+
+      this.optionTerms = [
+        this.underlying,
+        this.strikeAsset,
+        this.collateralAsset,
+        this.expiry,
+        this.strikePrice,
+        this.optionType,
+        this.paymentToken,
+      ];
+
+      //await this.adapter.populateOTokenMappings();
+
+      // console.log("this.underlying is %s", this.underlying);
+      // console.log("this.strikeAsset is %s", this.strikeAsset);
+      // console.log("this.collateralAsset is %s", this.collateralAsset);
+      // console.log("this.expiry is %s", this.expiry);
+      // console.log("this.strikeIndex is %s", this.strikeIndex);
+      // console.log("this.market is %s", this.market);
+      // console.log("isLong is %s", this.collateralAsset == ONE_ADDRESS ? true : false);
+      // console.log("this.strikePrice is %s", this.strikePrice);
+      // console.log("this.optionType is %s", this.optionType);
+      // console.log("this.paymentToken is %s", this.paymentToken);
+      // console.log("this.exerciseProfit is %s", this.exerciseProfit);
+      // console.log("this.purchaseAmount is %s", this.purchaseAmount);
+      // console.log("this.maxCost is %s", this.maxCost);
+      // console.log("user is %s", user);
+      // console.log("this.premium is %s", this.premium);
+    });
+
+    describe("#premium", () => {
+      it("gets premium of option", async function () {
+        assert.equal(
+          (await this.adapter.premium(this.optionTerms, this.purchaseAmount)).toString(),
+          this.premium
+        );
+      });
+    });
+
+    describe("#exerciseProfit", () => {
+      let snapshotId;
+
+      beforeEach(async () => {
+        snapshotId = await time.takeSnapshot();
+      });
+
+      afterEach(async () => {
+        await time.revertToSnapShot(snapshotId);
+      });
```

Comment:
We can DRY this up across the test file with:

```js
time.revertToSnapshotAfterEach();
```

Reply:
Example with no `beforeEach`: https://github.com/ribbon-finance/structured-products/blob/master/test/RibbonOptionsVault.js#L119

Example with `beforeEach`: https://github.com/ribbon-finance/structured-products/blob/master/test/RibbonOptionsVault.js#L154